File: bindings/wasm/tests/sol-vectors.test.ts
// Phase 6 Solana vector parity tests for the WASM binding.
//
// Mirrors `bindings/kotlin/.../SolVectorsTest.kt`.  Filters on `chain.kind`
// == "solana" / `unsigned_tx.kind` == "solana" / `message.kind` == "solana"
// for the positive-path tests, and `id.startsWith("sol.")` for the error
// vectors.

import { describe, it, expect } from 'vitest';

import { JovaWallet, JovaException } from '../src/index.js';
import vectors from '../../../spec/test-vectors.json';
import { decodeUnsigned, decodeSignableMessage } from './decoders.js';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const all: any[] = (vectors as any).vectors;

describe('SOL address vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'address') continue;
        if (v.input.chain.kind !== 'solana') continue;

        it(`address: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const got = wallet.address({ kind: 'solana' }, 0);
                expect(got.value).toBe(v.expected.address);
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one SOL address vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('SOL sign_tx vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'sign_tx') continue;
        if (v.input.unsigned_tx.kind !== 'solana') continue;

        it(`sign_tx: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const unsigned = decodeUnsigned(v.input.unsigned_tx);
                const signed = wallet.signTx(unsigned);
                expect(signed.raw_hex).toBe(v.expected.signed_hex);
                expect(signed.tx_hash).toBe(v.expected.tx_hash);
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one SOL sign_tx vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('SOL sign_message vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'sign_message') continue;
        if (v.input.message.kind !== 'solana') continue;

        it(`sign_message: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                const msg = decodeSignableMessage(v.input.message);
                const sig = wallet.signMessage(msg);
                // SOL convention reuses the `signature_hex` field for the
                // base58-encoded sig (Phase 1 cross-phase convention).
                expect(sig.hex).toBe(v.expected.signature_hex);
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one SOL sign_message vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

describe('SOL error vectors', () => {
    let ran = 0;
    for (const v of all) {
        if (v.kind !== 'error') continue;
        if (!String(v.id).startsWith('sol.')) continue;

        it(`error: ${v.id}`, () => {
            const wallet = JovaWallet.fromMnemonic({
                words: v.input.mnemonic,
                passphrase: v.input.passphrase ?? '',
            });
            try {
                let caught: unknown = null;
                try {
                    if (v.input.unsigned_tx) {
                        wallet.signTx(decodeUnsigned(v.input.unsigned_tx));
                    } else if (v.input.message) {
                        wallet.signMessage(decodeSignableMessage(v.input.message));
                    } else {
                        throw new Error(
                            `SOL error vector ${v.id} has neither unsigned_tx nor message`,
                        );
                    }
                } catch (e) {
                    caught = e;
                }
                expect(caught).toBeInstanceOf(JovaException);
                const expected = v.expected.error_variant as string;
                const expectedKind = expected.charAt(0).toLowerCase() + expected.slice(1);
                const payload = (caught as JovaException).error;
                expect(payload.kind).toBe(expectedKind);
                // For SOL errors, reason field is asserted (matches Kotlin
                // SolVectorsTest expectations).
                if ('reason' in payload && v.expected.reason) {
                    expect(payload.reason).toContain(v.expected.reason);
                }
            } finally {
                wallet.destroy();
            }
            ran++;
        });
    }

    it('exercised at least one SOL error vector', () => {
        expect(ran).toBeGreaterThan(0);
    });
});

File: bindings/kotlin/jova-core/src/test/kotlin/io/jova/core/SolVectorsTest.kt
package io.jova.core

import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.json.JSONObject
import org.json.JSONArray
import uniffi.jova_core_ffi.FfiException
import uniffi.jova_core_ffi.JovaChain
import uniffi.jova_core_ffi.JovaWallet
import uniffi.jova_core_ffi.SignableMessage
import uniffi.jova_core_ffi.UnsignedTx

/**
 * Phase 3c Solana vector parity tests.
 *
 * Loads `spec/test-vectors.json` (copied into the JVM test resources by
 * `bindings/kotlin/scripts/build-aar.sh`) and iterates every `sol.*` vector,
 * exercising the Kotlin uniffi binding through the same paths the Rust core
 * already verifies in `crates/jova-core/tests/vectors_sol.rs`.
 *
 * Mirrors `XrpVectorsTest`. Includes both transaction and message coverage,
 * plus error vectors that exercise MalformedUnsignedTx and
 * MalformedSignableMessage variants.
 *
 * Filters narrowly on `solana` / `sol.` discriminators so SOL vectors don't
 * leak into the BTC/EVM/XRP test runs (and vice-versa).
 */
class SolVectorsTest {

    private fun loadVectors(): JSONArray {
        val raw = javaClass.getResourceAsStream("/test-vectors.json")!!
            .bufferedReader().readText()
        return JSONObject(raw).getJSONArray("vectors")
    }

    @Test
    fun solAddressVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "address") continue
            val input = v.getJSONObject("input")
            if (input.getJSONObject("chain").getString("kind") != "solana") continue

            val id = v.getString("id")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val expected = v.getJSONObject("expected").getString("address")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val got = wallet.address(JovaChain.Solana, 0u)
            assertEquals(
                "SOL address mismatch for vector $id",
                expected,
                got.value
            )
            ran++
        }
        assertTrue("No SOL address vectors ran — check vector file", ran >= 1)
    }

    @Test
    fun solSignTxVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "sign_tx") continue
            val input = v.getJSONObject("input")
            val unsignedDict = input.getJSONObject("unsigned_tx")
            if (unsignedDict.getString("kind") != "solana") continue

            val id = v.getString("id")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val unsigned: UnsignedTx = decodeUnsignedTx(unsignedDict)
            val expectedHex = v.getJSONObject("expected").getString("signed_hex")
            val expectedHash = v.getJSONObject("expected").getString("tx_hash")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val signed = wallet.signTx(unsigned, 0u)

            assertEquals("signed_hex mismatch for vector $id", expectedHex, signed.rawHex)
            assertEquals("tx_hash mismatch for vector $id", expectedHash, signed.txHash)
            ran++
        }
        assertTrue("No SOL sign_tx vectors ran — check vector file", ran >= 2)
    }

    @Test
    fun solSignMessageVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "sign_message") continue
            val input = v.getJSONObject("input")
            val msgDict = input.getJSONObject("message")
            if (msgDict.getString("kind") != "solana") continue

            val id = v.getString("id")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val msg: SignableMessage = decodeSignableMessage(msgDict)
            // signature_hex carries the base58 string per the cross-phase
            // convention of using `signature_hex` regardless of encoding.
            val expected = v.getJSONObject("expected").getString("signature_hex")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            val sig = wallet.signMessage(msg, 0u)

            assertEquals("SOL signature mismatch for vector $id", expected, sig.hex)
            ran++
        }
        assertTrue("No SOL sign_message vectors ran — check vector file", ran >= 1)
    }

    @Test
    fun solErrorVectors() {
        val vectors = loadVectors()
        var ran = 0
        for (i in 0 until vectors.length()) {
            val v = vectors.getJSONObject(i)
            if (v.getString("kind") != "error") continue
            val id = v.getString("id")
            if (!id.startsWith("sol.")) continue

            val input = v.getJSONObject("input")
            val mnemonic = input.getString("mnemonic")
            val pass = if (input.has("passphrase")) input.getString("passphrase") else ""
            val expectedVariant = v.getJSONObject("expected").getString("error_variant")
            val expectedReason = v.getJSONObject("expected").getString("reason")

            val wallet = JovaWallet.fromMnemonic(mnemonic, pass)
            try {
                when {
                    input.has("unsigned_tx") -> {
                        val unsigned = decodeUnsignedTx(input.getJSONObject("unsigned_tx"))
                        wallet.signTx(unsigned, 0u)
                    }
                    input.has("message") -> {
                        val msg = decodeSignableMessage(input.getJSONObject("message"))
                        wallet.signMessage(msg, 0u)
                    }
                    else -> fail("SOL error vector $id must carry an unsigned_tx or message in input")
                }
                fail("Expected FfiException.$expectedVariant for vector $id, but no exception was thrown")
            } catch (err: FfiException) {
                val actualVariant = err::class.simpleName ?: "Unknown"
                assertEquals(
                    "Wrong error variant for vector $id",
                    expectedVariant,
                    actualVariant
                )
                val msg = err.message ?: ""
                assertTrue(
                    "Error message for $id should contain reason '$expectedReason', got: $msg",
                    msg.contains(expectedReason)
                )
                ran++
            }
        }
        assertTrue("No SOL error vectors ran — check vector file", ran >= 4)
    }
}
